File: DesignPatterns/01_GOF23/DP_examples/Go/01_Creational/Abstract Factory/AbstractFactory.go
package AbstractFactory

type AbstractFactory interface {
	GetShape() IShape
	GetColor() IColor
}

type productFactory[Color IColor, Shape IShape] struct{}

func (p productFactory[Color, Shape]) GetShape() IShape {
	return *new(Shape)
}
func (p productFactory[Color, Shape]) GetColor() IColor {
	return *new(Color)
}

File: DesignPatterns/01_GOF23/DP_examples/Rust/01_Creational/Abstract Factory/abstract_factory.rs
// abstract_factory.rs
// 抽象工厂

pub trait AbstractFactory {
    fn get_color(&self) -> Box<dyn Color>;
    fn get_shape(&self) -> Box<dyn Shape>;
}

pub trait Color {
    fn fill(&self);
}

pub trait Shape {
    fn draw(&self);
}
